### ISSUE TITLE
waagent upgrade and downloaded all the extensions and enable again

### ISSUE THREAD

**Issue by sundaxi** [Azure/WALinuxAgent#1067]
I noticed that when waagent upgrade, it will try to download all the exsiting extension again and enable them all. But sometimes, some extension cannot be enabled correctly. And the vm state kept in seeking state forever. 

My question is
1. What's the timeout for enable a non-work extension? It seems that the fabric controller will wait for the update from extension status forever. 
2. Can we set timeout timer manually? 
2. For some one time running extension like VMAaccess and custom script, why they need to be re-installed every time when agent upgraded?

**Comment by boumenot** [comment]
1. The timeout is 15 minutes for install, and 5 minutes for enable.
 1. No.  You could modify the code, and re-deploy the agent. 😄 
 1. Extensions should be re-installed upon an agent upgrade.  Extensions are enabled whenever the agent is restarted.

**Comment by sundaxi** [comment]
Thanks @boumenot . Whenthe timeout triggered for extension enable failure. For my understanding the waagent should send out a report to wireserver to let it know the state failed and release the seeking goal state from Azure fabric controller. But in many cases, it doesn't report wireserver and Fabric controller mark the state of vm as seeking forever. Is there any way to avoid to fix such kind of scenario ?

**Comment by boumenot** [comment]
The agent uploads the latest status file for an extension to the VM's status blob.  It does not every few seconds.  WireServer is not used for this.  The status blob is a special blob in Azure blob storage.

If the agent calls an operation on an extension, such as Install or Enable it will issue a telemetry event.  It sends telemetry events via WireServer.  Regardless of an operation though, the agent continually uploads the extension's status.

**Comment by sundaxi** [comment]
thanks for the explanation of how does it works. 
I met with one issue that the agent upgraded and installed the exsiting extension again but failed, see logs below. But agent never report it to wireserver, and the Fabric controller marked this vm as goal state "seeking" forever. I have to uninstall this extension to migrate the issue. Can you please advise :)

2018/03/14 08:57:55.319184 INFO [Microsoft.Compute.CustomScriptExtension-1.9] Target handler state: enabled
2018/03/14 08:57:55.321912 INFO [Microsoft.Compute.CustomScriptExtension-1.9] [Enable] current handler state is: notinstalled
2018/03/14 08:57:55.481553 INFO Event: name=Microsoft.Compute.CustomScriptExtension, op=Download, message=Download succeeded, duration=0
2018/03/14 08:57:55.483676 INFO [Microsoft.Compute.CustomScriptExtension-1.9] Initialize extension directory
2018/03/14 08:57:55.485531 INFO [Microsoft.Compute.CustomScriptExtension-1.9] Update settings file: 0.settings
2018/03/14 08:57:55.485962 INFO [Microsoft.Compute.CustomScriptExtension-1.9] Install extension [install.cmd]
2018/03/14 08:57:56.501024 ERROR Event: name=Microsoft.Compute.CustomScriptExtension, op=Install, message=[000003] Non-zero exit code: 127, install.cmd, duration=0

**Comment by boumenot** [comment]
It looks like someone tried to install a Windows extension on a Linux VM.  The custom script extension for Linux is called Microsoft.Azure.Extensions.CustomScript.  The custom script extension for Window is called Microsoft.Compute.CustomScriptExtension.

**Comment by thefueley** [comment]
Anyone know why my command just hangs in PowerShell when I issue the following: 

$customConfig = @{"commandToExecute" = "scp user@server:/home/user/motd.sh .; chmod +x motd; /bin/bash motd.sh"};

$vmss = Get-AzureRmVmss `
          -ResourceGroupName "myRG" `
          -VMScaleSetName "mySS"


$vmss = Add-AzureRmVmssExtension `
  -VirtualMachineScaleSet $vmss `
  -Name "customScript" `
  -Publisher "Microsoft.Azure.Extensions `
  -Type "CustomScript" `
  -TypeHandlerVersion 2.0.2 `
  -Setting $customConfig


Update-AzureRmVmss `
  -ResourceGroupName "myRG" `
  -Name "mySS" `
  -VirtualMachineScaleSet $vmss

It just hangs, waiting for more input or something after the Add-AzureRmVmssExtension command.

I tried it with TypeHandlerVersion 1.8 and also 2.0.2. I got that from the output of Get-AzureRmVmExtensionImage.
